### PR TITLE
Add secondary navigation

### LINKED ISSUE #182: Secondary navigation
As well as primary navigation, some services have secondary navigation as second horizontal row of navigation options.

Here’s an example from Manage training for early career teachers:

Screenshot 2023-12-12 at 10 07 28

The style there is very similar to the primary navigation, except with no background colour, and the current item has black text and a blue border at the bottom. It's placed after an h1 which is the scope for the secondary nav, and then the current item is repeated as an h2 after (possibly this isn't needed). 

Arguably this is similar to the [Tabs component](https://design-system.service.gov.uk/components/tabs/), but using links and full pages instead of javascript.

Worth considering adding as a prototype component?

### PR DESCRIPTION
**Pull request by frankieroberto**
This adds a secondary navigation component, plus some documentation about it.

Resolves #182

### CODE REVIEW

**Comment by frankieroberto** [comment]
This needs fixing... 😞 

Screenshot 2023-12-14 at 11 28 43

File: x-govuk/components/secondary-navigation/_secondary-navigation.scss [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    border-left-color: transparent;
+    position: relative;
+  }
+}
+
+.x-govuk-secondary-navigation__list-item--current {
+  a,
+  a:link,
+  a:visited {
+    border-color: $govuk-brand-colour;
```
Comment (paulrobertlloyd) [inline x-govuk/components/secondary-navigation/_secondary-navigation.scss:77]: Should this be link colour like in the primary navigation component? Although I note there’s an open issue on https://github.com/govuk-one-login/service-header/issues/29 suggesting that the current item border uses the brand colour. 

It’s hard to know the intention for many things in the GOV.UK Design System given that the brand colour is also the same as the link colour. I think I’m inclined to go towards keeping this as using the brand colour, and updating the primary navigation to match.

File: x-govuk/components/secondary-navigation/_secondary-navigation.scss [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    color: $govuk-text-colour;
+  }
+
+  a:focus {
+    color: $govuk-focus-text-colour;
+    background-color: $govuk-focus-colour;
+    box-shadow: 0 -2px $govuk-focus-colour, 0 5px $govuk-focus-text-colour;
+    border-color: $govuk-focus-text-colour;
+    border-left-color: transparent;
```
Comment (paulrobertlloyd) [inline x-govuk/components/secondary-navigation/_secondary-navigation.scss:86]: I’d be tempted to keep this using the link focus style here, and for a number of reasons:

1. In GOV.UK Frontend, there’s discussion around having many different focus styles (weighed up against ensuring each focus style is distinct enough from the unfocused state for a given component). This is yet another focus style.

2. It’s visually inconsistent with the primary navigation focus style:
  
    Screenshot of focus state in primary navigation

3. It adds additional styles where we can lean on either existing CSS via the cascade or using existing link `@mixins`

File: x-govuk/components/secondary-navigation/template.njk [hunk 3]
```diff
@@ -0,0 +1,14 @@
+<nav class="x-govuk-secondary-navigation {%- if params.classes %} {{ params.classes }}{% endif -%}" aria-label="Sub navigation" {%- for attribute, value in params.attributes %} {{ attribute }}="{{ value }}"{% endfor %}>
+  <ul class="x-govuk-secondary-navigation__list">
+    {% for item in params.items %}
+      <li class="x-govuk-secondary-navigation__list-item {{ "x-govuk-secondary-navigation__list-item--current" if item.current }}{% if item.classes %} {{ item.classes }}{% endif -%}">
+        <a class="govuk-link x-govuk-secondary-navigation__link govuk-link--no-visited-state" href="{{ item.href }}" {{ "aria-current=page" if item.current }}>
+          {{ item.text }}
+          {% if item.current %}
+            <span class="govuk-visually-hidden">Current page</span>
```
Comment (paulrobertlloyd) [inline x-govuk/components/secondary-navigation/template.njk:8]: Is this needed? Wouldn’t the presence of `aria-current` mean this state would be announced twice?

File: docs/examples/secondary-navigation-with-header.njk [hunk 4]
```diff
@@ -0,0 +1,29 @@
+---
+eleventyExcludeFromCollections: true
+layout: example.njk
+title: Secondary navigation with header
+---
+{% from "x-govuk/components/secondary-navigation/macro.njk" import xGovukSecondaryNavigation %}
+
+
```
Comment (paulrobertlloyd) [inline docs/examples/secondary-navigation-with-header.njk:8]:
```suggestion

```

File: docs/secondary-navigation.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+{{ appExample({
+  component: "secondary-navigation",
+  example: "with-header"
+}) }}
+
+## When not to use this component
+
+Do not use this component for the primary sections of your service, or if a [sub-navigation](sub-navigation) would be more appropriate.
+
+You could also consider using the [tabs component](https://design-system.service.gov.uk/components/tabs/) if the amount of content in each sub-section is small.
```
Comment (paulrobertlloyd) [inline docs/secondary-navigation.md:29]: Do we want to add something about how tabs should be used client-side, and secondary navigation should be used server-side, if that makes sense?

Maybe unnecessary as this is the default behaviour, but I’ve seen cases in the past where the tab component has been ‘hijacked’ to link to different pages, whereas this component should be used in that instance instead.

**Review by paulrobertlloyd — COMMENTED** [review]
Initial thoughts from a quick look.

File: x-govuk/components/secondary-navigation/_secondary-navigation.scss (hunk 2 above)
Comment (frankieroberto) [reply x-govuk/components/secondary-navigation/_secondary-navigation.scss:86]: I'm not sure about the focus and hover states either. For the primary nav, the trend seems to be moving towards the focus area being bigger, making it the same size as the clickable area, rather than smaller.

Eg the [Design System website](https://design-system.service.gov.uk) uses the smaller focus area, but the [One Login header](https://govuk-one-login.github.io/service-header/dist/preview.html) uses the larger one, as does the [MOJ primary navigation](https://design-patterns.service.justice.gov.uk/components/primary-navigation/).

Personally, I think I prefer the larger focus area, mainly to avoid the double-border below the text? 🤷 

Agree that the primary and secondary nav should be consistent though...

File: x-govuk/components/secondary-navigation/_secondary-navigation.scss (hunk 1 above)
Comment (frankieroberto) [reply x-govuk/components/secondary-navigation/_secondary-navigation.scss:77]: Agree, feels like this should be the brand colour.

File: x-govuk/components/secondary-navigation/template.njk (hunk 3 above)
Comment (frankieroberto) [reply x-govuk/components/secondary-navigation/template.njk:8]: Not sure either, but it was already in the code I copied... 😂 

Happy to drop this, unless anyone can remember any good reasons why it was first added...

File: docs/secondary-navigation.md (hunk 5 above)
Comment (frankieroberto) [reply docs/secondary-navigation.md:29]: I'm not sure we need to use the words client-side or server-side, but maybe we could go into a bit more detail about the difference between navigation and tabs, and when to use (and not use) either one?

The Tabs page goes into a fair bit of detail, including this:

> Do not use the tabs component if the total amount of content the tabs contain will make the page slow to load. For this reason, do not use the tabs component as a form of page navigation.

But yeah, I’ve also seen this ignored and the Tabs styled re-used as navigation instead (which is probably good evidence for the need for secondary navigation?).

File: docs/secondary-navigation.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+## When to use this component
+
+Use the secondary navigation when you already have a [primary navigation](primary-navigation) and need to let users navigate between pages that are scoped to a particular section or concept.
+
+For example, if your service is a case management system, you might use the secondary navigation if there are multiple pages associated with a single case.
+
+{{ appExample({
+  component: "secondary-navigation",
+  example: "with-header"
+}) }}
```
Comment (frankieroberto) [inline docs/secondary-navigation.md:23]: @paulrobertlloyd  I added this second example as I thought for this component it was helpful to see the component within a bit more context (ie with a heading above and below it) to help distinguish it from the primary nav.  But that seemed too much for the preview at the very top of the page...

Comment (paulrobertlloyd) [reply docs/secondary-navigation.md:23]: Makes sense to me.

File: x-govuk/components/secondary-navigation/_secondary-navigation.scss (hunk 2 above)
Comment (paulrobertlloyd) [reply x-govuk/components/secondary-navigation/_secondary-navigation.scss]: The problem with the bigger focus area is that it’s not distinct from a currently selected item. For example, removing all colour, a focused item looks like a currently highlighted item:

* GOV.UK Design System navigation (current versus focused):

    Screenshot of navigation from GOV.UK Design System website.

* GOV.UK One Login service header example (focused versus current):

    Screenshot of navigation from GOV.UK One Login service header.

I note that you increased the bottom border on a focused area which avoid this issue, but it introduces a whole new focus style. Perhaps something to file an issue on upstream on the One login repo?

For now, and for the sake of simplicity and following current convention, perhaps we use the link focus styles in the secondary navigation?

File: docs/secondary-navigation.md (hunk 5 above)
Comment (paulrobertlloyd) [reply docs/secondary-navigation.md:29]: Wasn’t suggesting actually saying client-side/sever-side 😉 Perhaps something we can ignore until we see a need to add additional guidance?

File: x-govuk/components/secondary-navigation/template.njk (hunk 3 above)
Comment (paulrobertlloyd) [reply x-govuk/components/secondary-navigation/template.njk:8]: Where did you copy the code from?

**Comment by paulrobertlloyd** [comment]
Have fixed the homepage layout, and rebased from the `main` branch where I updated linting on the project.

File: x-govuk/components/secondary-navigation/template.njk [hunk 7]
```diff
@@ -0,0 +1,14 @@
+<nav class="x-govuk-secondary-navigation {%- if params.classes %} {{ params.classes }}{% endif -%}" aria-label="Sub navigation" {%- for attribute, value in params.attributes %} {{ attribute }}="{{ value }}"{% endfor %}>
```
Comment (paulrobertlloyd) [inline x-govuk/components/secondary-navigation/template.njk:1]: We should follow the example of the primary navigation component and label the navigation with a header. Or maybe we stick with `aria-label` in this case, but the default value should be something akin to `Sub sections`, but either way it shouldn’t include the word ‘navigation’ as that’s implicit and will be announced already because of the navigation role.

However, the example you include in the guidance is a good example where you’d want to have the navigation labelled by another heading or element on the page already (i.e. ‘James Smith’). Maybe we provide an option to override the value for `aria-labelledby`, and if that’s given we don’t add the visually hidden heading?

So, with all that in mind, something like this: 

```suggestion
<nav class="x-govuk-secondary-navigation {%- if params.classes %} {{ params.classes }}{% endif -%}" aria-labelledby="{{ params.labelledby or "secondary-navigation-heading" }}"{%- for attribute, value in params.attributes %} {{ attribute }}="{{ value }}"{% endfor %}>
{% if not params.labelledby %}
  <h2 class="govuk-visually-hidden" id="secondary-navigation-heading">{{ params.visuallyHiddenTitle or "Sub sections" }}</h2>
{% endif %}
```

File: x-govuk/components/secondary-navigation/template.njk (hunk 3 above)
Comment (frankieroberto) [reply x-govuk/components/secondary-navigation/template.njk]: I can't actually remember when I copied the code from 😬  - I was trying to quick replicate the version we have in production.

Digging through our production git history, it looks like this [was copied](https://github.com/DFE-Digital/early-careers-framework/pull/274/commits/a41842ca4ffef0495724c4cf9b5f3d96ba7a5db6) (2 years ago) from [the version in the HMCTS design system](https://github.com/hmcts/design-system/blob/master/app/components/subnav/_subnav.scss).

Separately, a bunch of other DfE services have a very visually-similar design which uses a different implementation [originally added](https://github.com/DFE-Digital/manage-teacher-training-applications-prototype/pull/196) by @adamsilver (after a crit from you @paulrobertlloyd!).

All of which is good evidence to consolidate all this into a shared component, which can eventually make its way into the GOV.UK Design System 🤞.

Anyway, I've removed the visually-hidden "Current page" text now in 91f2373746c044257f956ba64158ff06746aa429, on the assumption that this is redundant given we’ve added `aria-current="page"` like in the primary nav.

File: x-govuk/components/secondary-navigation/template.njk (hunk 7 above)
Comment (frankieroberto) [reply x-govuk/components/secondary-navigation/template.njk]: I've updated this in 7cbcd1f1ca333bd8f3fab4b00da3fb68ee231747 so that the `aria-label` text is set by `params.visuallyHiddenTitle` for consistency with the primary navigation.

Being able to set `aria-labelledby` value for an ID which is outside of the component itself is interesting - but feels error-prone?   I'm not enough of an accessibility expert to know whether it's better to use an `aria-label` or a reference to an existing heading.  Probably the whole thing needs a bit of an accessibility review.

File: x-govuk/components/secondary-navigation/_secondary-navigation.scss (hunk 2 above)
Comment (frankieroberto) [reply x-govuk/components/secondary-navigation/_secondary-navigation.scss]: I've done a big refactor in 26adff173b14d8c85f3ab1e7faaa58c8e8b66b91 which reverts the css to be more similar to the primary navigation, with the focus highlight only extending to the text, and the underline on hover also being just below the text rather than where the selected item border is.

I do still think that having the double-border is a bit visually-jarring, but I agree we should keep the secondary nav consistent with the primary one, and then if there's any changes we could apply them to both.

File: docs/secondary-navigation.md (hunk 5 above)
Comment (frankieroberto) [reply docs/secondary-navigation.md]: Update the guidance - take another look and let me know what you think. I added a "How it works" section, following the convention from the GOV.UK Design System website.

**Comment by frankieroberto** [comment]
@paulrobertlloyd I've reworked the CSS on this (rather than the copy-and-paste job I had before), so now it's more similar to the primary navigation component implementation.

This means the same hover and focus styles are used as primary nav - ie with the double underline on hover, and the focus highlight being smaller.  We can always review this separately if thinking on this changes.

This also means that the specific mobile view has been dropped, in favour of doing the same as the primary nav - ie horizontal with wrapping.

Let me know what you think!

File: docs/_components/example/template.njk [hunk 8]
```diff
@@ -8,6 +8,9 @@
 {%- endset -%}
 
 {%- set examplePath = "/examples/" + params.component -%}
+{% if params.example %}
+  {%- set examplePath = examplePath + "-" + params.example -%}
+{% endif %}
```
Comment (paulrobertlloyd) [inline docs/_components/example/template.njk:13]: In #192 I wanted to show the full code example in the Nunjucks tab, not the component isolated on its own. To achieve that, I’ve take the bit from this PR where each tab gets a unique identifier, but have avoided splitting up `component` and `example` into separate params meaning each example shows the preview and the associated code used to create it. 

If that makes sense. Can update this PR if it doesn’t!

**Comment by frankieroberto** [comment]
@paulrobertlloyd I've tweaked the padding so that it's reduced on mobile, which I think helps to make it clearer which item is the current page?

localhost_8080_examples_secondary-navigation_(iPhone 5_SE) (1)

In reality, I think for most services, the nav should fix on a single line for most mobile viewports (given how much wider phones are nowadays).

We could still consider an option to have the items stack vertically on mobile (with or without a toggle) - but I think if we do that, it's worth doing it for the primary nav and secondary nav at the same time?

File: docs/_components/example/template.njk (hunk 8 above)
Comment (frankieroberto) [reply docs/_components/example/template.njk:13]: Not quite sure I follow this. I added the 'example' option as a quick way to have multiple examples per component, based loosely on the design system one (which has 3 options: `group`, `item` and `example`).

Maybe that's overkill though, and we could just have a single param named `example` (instead of `component`) for now, all within a single folder?

Comment (paulrobertlloyd) [reply docs/_components/example/template.njk:13]: Yeah, let’s keep it simple for now and rename the param `example`. We can split this up later, if we need to.

File: x-govuk/components/secondary-navigation/template.njk [hunk 9]
```diff
@@ -0,0 +1,11 @@
+<nav class="x-govuk-secondary-navigation {%- if params.classes %} {{ params.classes }}{% endif -%}" aria-label="{{ params.visuallyHiddenTitle or "Secondary Menu" }}" {%- for attribute, value in params.attributes %} {{ attribute }}="{{ value }}"{% endfor %}>
```
Comment (paulrobertlloyd) [inline x-govuk/components/secondary-navigation/template.njk:1]: What about `aria-labelled` by? I think that’d work well for the second example used in the documentation for this component.

File: docs/secondary-navigation.md [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+One of the items should always be highlighted as the current page. This item remains a link, but uses black text and has a line beneath it.
+
+The secondary navigation should appear before the `<main>` element of the page. This means that it is skipped when [skip link](https://design-system.service.gov.uk/components/skip-link/) is used.
+
+The title of the current page can be repeated below the sub navigation, inside the `<main>` element, using a heading tag.
+
+You can use HTML or, if you are using Nunjucks or the GOV.UK Prototype Kit, you can use the Nunjucks macro.
+
+### Nunjucks macro options
```
Comment (paulrobertlloyd) [inline docs/secondary-navigation.md:43]: Tempted to update documentation for the other components to use this heading. Can do so in a separate PR. 

```suggestion
### Nunjucks macro options
```

File: docs/examples/secondary-navigation.njk [hunk 11]
```diff
@@ -0,0 +1,23 @@
+---
+eleventyExcludeFromCollections: true
+layout: example.njk
+title: Secondary navigation
+---
+{% from "x-govuk/components/secondary-navigation/macro.njk" import xGovukSecondaryNavigation %}
+
+{{ xGovukSecondaryNavigation({
```
Comment (paulrobertlloyd) [inline docs/examples/secondary-navigation.njk:8]: Prevent leading new line in Nunjucks code preview:

```suggestion
{% from "x-govuk/components/secondary-navigation/macro.njk" import xGovukSecondaryNavigation %}
{{ xGovukSecondaryNavigation({
```

**Review by paulrobertlloyd — CHANGES_REQUESTED** [review]
Let’s leave thoughts of responsive versions of this for now; as you say, in most cases shouldn’t be needed and the items wrap much better now.

Also… do you think an example layout for the Prototype Kit would be a useful addition?

File: docs/_components/example/template.njk (hunk 8 above)
Comment (frankieroberto) [reply docs/_components/example/template.njk:13]: Ok, done in #196

File: docs/secondary-navigation.md (hunk 10 above)
Comment (frankieroberto) [reply docs/secondary-navigation.md:42]: Yeah, I thought it was worth following the term used on the GOV.UK Design System website (although there it's hidden away behind a Details, which I often forget about).

File: x-govuk/components/secondary-navigation/_secondary-navigation.scss [hunk 12]
```diff
@@ -0,0 +1,64 @@
+.x-govuk-secondary-navigation {
```
[the remark this answers is not in the document]
Comment (frankieroberto) [reply x-govuk/components/secondary-navigation/_secondary-navigation.scss:1]: ok. Is this something a linter could auto-fix? (I am lazy/forgetful with this stuff)

File: x-govuk/components/secondary-navigation/_secondary-navigation.scss [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+  @include govuk-clearfix;
+  list-style: none;
+  margin: 0;
+  padding: 0;
+  width: 100%;
+
+  // The list uses box-shadow rather than a border to set a 1px
+  // grey line at the bottom, so that border from the current
+  // item appears on top of the grey line.
+  box-shadow: inset 0 -1px 0 $govuk-border-colour;
```
[the remark this answers is not in the document]
Comment (frankieroberto) [reply x-govuk/components/secondary-navigation/_secondary-navigation.scss:31]: Dunno - feels like this line is mostly aesthetic and non-critical, so long as the line underneath the current item always appears even in high contrast mode? 🤷‍♂️

File: docs/secondary-navigation.md [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+
+| Name                    | Type   | Description                                                                                                                  |
+| :---------------------- | :----- | :--------------------------------------------------------------------------------------------------------------------------- |
+| **items**               | array  | **Required**. An array of navigation links within the secondary navigation. See [items](#options-for-items).                 |
+| **labelledBy**          | string | A reference to the id of an existing heading on a page which actuals as the label for the navigation                         |
```
Comment (paulrobertlloyd) [inline docs/secondary-navigation.md:49]:
```suggestion
| **labelledBy**          | string | A reference to the id of an existing heading on a page which acts as the label for the navigation.                         |
```

File: docs/secondary-navigation.md [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+| **visuallyHiddenTitle** | text   | A visually-hidden label to describe the navigation. Defaults to ‘Secondary Menu’. Not included if labelledBy is set instead. |
```
Comment (paulrobertlloyd) [inline docs/secondary-navigation.md:50]:
```suggestion
| **visuallyHiddenTitle** | text   | A visually-hidden label to describe the navigation. Defaults to ‘Secondary Menu’. Not included if `labelledBy` is set instead. |
```

File: docs/secondary-navigation.md [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+| Name        | Type    | Description                                                        |
+| :---------- | :------ | :----------------------------------------------------------------- |
+| **text**    | string  | **Required**. Text of the navigation link.                         |
+| **href**    | array   | **Required**. The value of the navigation link’s `href` attribute. |
+| **current** | boolean | Indicate that the item is the current page.                        |
+| **classes** | string  | Classes to add to the navigation item.                             |
+
+## Research on this component
+
+This component is based on ones originally developed as part of the MOJ and HMCTS design systems.
```
Comment (paulrobertlloyd) [inline docs/secondary-navigation.md:65]: Avoid abbreviations where possible:

```suggestion
This component is based on ones originally developed as part of the Ministry of Justice and HMCTS design systems.
```

File: docs/secondary-navigation.md [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+  example: "secondary-navigation-with-header"
+}) }}
+
+Unlike [tabs](https://design-system.service.gov.uk/components/tabs/), secondary navigation does not use JavaScript. Each item within the navigation is a link to a separate full page.
+
+One of the items should always be highlighted as the current page. This item remains a link, but uses black text and has a line beneath it.
+
+The secondary navigation should appear before the `<main>` element of the page. This means that it is skipped when [skip link](https://design-system.service.gov.uk/components/skip-link/) is used.
+
+The scope of the secondary navigation can be given by a heading above it. When doing this, you should set `labelledBy` to be the id of the heading, so that it is associated with the heading for screen reader users.
```
Comment (paulrobertlloyd) [inline docs/secondary-navigation.md:38]: Avoid using description of visual placement (could be below, to one side…)

```suggestion
The scope of the secondary navigation can be given by an associated heading. When doing this, you should set `labelledBy` to be the `id` of the heading, so that it is associated with the heading for screen reader users.
```

File: x-govuk/components/secondary-navigation/_secondary-navigation.scss (hunk 12 above)
[the remark this answers is not in the document]
Comment (paulrobertlloyd) [reply x-govuk/components/secondary-navigation/_secondary-navigation.scss:1]: Er… no. It’s not included in the Stylelint config. Yet.

**Review by paulrobertlloyd — CHANGES_REQUESTED** [review]

File: x-govuk/components/secondary-navigation/_secondary-navigation.scss (hunk 13 above)
[the remark this answers is not in the document]
Comment (paulrobertlloyd) [reply x-govuk/components/secondary-navigation/_secondary-navigation.scss:31]: Maybe? I’m not sure what expectations are around high contrast. Easily something we can come back to.

File: x-govuk/components/secondary-navigation/_secondary-navigation.scss (hunk 12 above)
[the remark this answers is not in the document]
Comment (frankieroberto) [reply x-govuk/components/secondary-navigation/_secondary-navigation.scss:1]: Boo. Well I personally would ban any style rules unless they can be auto-enforced... 😉

**Review by paulrobertlloyd — APPROVED** [review]
